src/inmanta/agent/agent.py

```diff
@@ -192,24 +269,160 @@ async def _execute(self, ctx: handler.HandlerContext, requires: dict[ResourceIdS
+        else:
+            await self._execute(resource, ctx=ctx, requires=requires)
+
+        ctx.debug(
+            "End run for resource %(r_id)s in deploy %(deploy_id)s",
+            r_id=resource_id,
+            deploy_id=gid,
+        )
+
+        if ctx.facts:
+            ctx.debug("Sending facts to the server")
+            set_fact_response = await self.client.set_parameters(tid=env_id, parameters=ctx.facts)
+            if set_fact_response.code != 200:
+                ctx.error("Failed to send facts to the server %s", set_fact_response.result)
+
+
+        changes: dict[ResourceVersionIdStr, dict[str, AttributeStateChange]] = {resource_id.resource_version_str(): ctx.changes}
+        response = await self.client.resource_deploy_done(
+            tid=env_id,
+            rvid=resource_id.resource_version_str(),
+            action_id=ctx.action_id,
+            status=ctx.status,
+            messages=ctx.logs,
+            changes=changes,
+            change=ctx.change,
+        )
+        if response.code != 200:
+            LOGGER.error("Resource status update failed %s", response.result)
+
+        return ctx.status, ctx.change, changes
+
+
+class ExternalExecutor(Executor):
+    """
+    This class encapsulates an executor running in a separate process from the agent.
+    The AgentSupervisor is responsible for spawning and managing these executors
+    on demand of the ResourceScheduler.
+    """
+
+    async def execute(
+        self,
+        gid: uuid.UUID,
+        resource_id: Id,
+        resource: dict[str, object],
+        env_id: uuid.UUID,
+        ):
+        raise NotImplementedError
+
+    @classmethod
+    async def get_executor(cls, agent: "AgentInstance", code: Sequence[ResourceInstallSpec], client: protocol.Client):
+        raise NotImplementedError
+
+class ResourceAction(ResourceActionBase):
+    def __init__(
+        self,
+        scheduler: "ResourceScheduler",
+        executor: Executor,
+        env_id: uuid.UUID,
+        resource: dict[str, object],
+        gid: uuid.UUID,
+        reason: str,
+    ) -> None:
+        """
+        :param gid: A unique identifier to identify a deploy. This is local to this agent.
+        """
+        super().__init__(scheduler, Id.parse_id(resource["id"]), gid, reason)
+        self.resource: dict[str, object] = resource
+        self.executor: Executor = executor
+        self.env_id: uuid.UUID = env_id
+
     async def execute(
         self, dummy: "ResourceActionBase", generation: "Dict[ResourceIdStr, ResourceActionBase]", cache: AgentCache
     ) -> None:
         self.logger.log(const.LogLevel.TRACE.to_int, "Entering %s %s", self.gid, self.resource)
-        with cache.manager(self.resource.id.get_version()):
-            self.dependencies = [generation[x.resource_str()] for x in self.resource.requires]
+        with cache.manager(self.resource_id.get_version()):
```

Comment:
the cache will have to live on the executor. 

This will somewhat complicate the executor, as it will have state (the cache state)